File: inn/overrides/payment_entry.py
# /apps/inn/inn/overrides/payment_entry.py
# -*- coding: utf-8 -*-
# Copyright (c) 2020, Core Initiative and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, cstr

from inn.services.payment_entry_service import PaymentEntryService


# ---------------------------------------------------------------------
# Helper: determine canonical Payment Entry amount (safe across versions)
# ---------------------------------------------------------------------
def _get_payment_entry_effective_amount(pe_doc):
    """
    Determine canonical amount of a Payment Entry.
    Preference order:
      1) paid_amount
      2) received_amount
      3) base_received_amount
      4) base_paid_amount
      5) total_allocated_amount (fallback)
      6) 0.0
    """
    return flt(
        getattr(pe_doc, "paid_amount", None)
        or getattr(pe_doc, "received_amount", None)
        or getattr(pe_doc, "base_received_amount", None)
        or getattr(pe_doc, "base_paid_amount", None)
        or getattr(pe_doc, "total_allocated_amount", None)
        or 0.0
    )


# ---------------------------------------------------------------------
# API: used by Payment Entry UI to fetch unpaid ARIs for a customer
# ---------------------------------------------------------------------
@frappe.whitelist()
def get_unpaid_city_ledger_invoices(customer_id):
    """
    Return unpaid AR City Ledger Invoices for a given customer.
    """
    if not customer_id:
        frappe.throw(_("Please select a Customer first."))

    unpaid_invoices = frappe.get_all(
        "AR City Ledger Invoice",
        filters={
            "customer_id": customer_id,
            "status": ("!=", "Paid"),
        },
        fields=["name", "total_amount", "outstanding"],
        order_by="modified desc",
    )

    result = []
    for invoice in unpaid_invoices:
        result.append(
            {
                "reference_doctype": "AR City Ledger Invoice",
                "reference_name": invoice.name,
                "total_amount": flt(invoice.total_amount),
                "outstanding_amount": flt(invoice.outstanding),
                "allocated_amount": 0.0,
            }
        )

    return result


def validate_payment_entry_allocation_balance(doc, method):
    # Make sure that the Paid amount total is the sum of all allocated in the City ledger table + the payment table
    # Loop through the CLI and check if the allocated amount is equal to or lower that the outstanding amount.
    # If it is, then add the allocated amount to the paid amount total.
    # If it is not, then throw an error and do not submit the payment entry.
    # Add the payment to each of the CLI and update the paid amount total.
    # Update status of the CLI to paid if the paid amount total is equal to the total amount.
    payment_entry_service = PaymentEntryService(doc)
    payment_entry_service.validate_allocated_amount()
    payment_entry_service.allocate_city_ledger_invoices()


# ---------------------------------------------------------------------
# Hook: Payment Entry on_submit
# - Validate allocations provided in doc.custom_unpaid_city_ledger_invoice
# - Apply allocations to AR City Ledger Invoice child table and update totals
# ---------------------------------------------------------------------
def on_payment_entry_submit_custom_logic(doc, method):
    """Allocate payment entry to AR City Ledger Invoices using optimized service."""
    payment_entry_service = PaymentEntryService(doc)
    payment_entry_service.allocate_city_ledger_invoices()


# ---------------------------------------------------------------------
# Hook: Payment Entry on_cancel
# - Remove all rows in AR City Ledger Invoice that reference this PE
# - Update totals and status accordingly
# ---------------------------------------------------------------------
def on_payment_entry_cancel_custom_logic(doc, method):
    """
    Called on Payment Entry cancel to rollback allocations created at submit.
    """
    # Try to call centralized remover in ARCI module if exists
    try:
        remover = frappe.get_attr(
            "inn.inn_hotels.doctype.ar_city_ledger_invoice.ar_city_ledger_invoice.remove_payment_entry_links"
        )
        try:
            removed = remover(doc.name)
            # If remover returns a list, optionally log it
            if removed:
                frappe.msgprint(
                    _(
                        "Removed Payment Entry {0} allocations from AR City Ledger Invoices: {1}"
                    ).format(doc.name, ", ".join(removed))
                )
            return
        except Exception:
            # fallback to local removal if centralized function raises
            pass
    except Exception:
        # centralized remover not present, perform local fallback
        pass

    # Local fallback removal:
    try:
        rows = frappe.db.sql(
            """
            SELECT parent, name, payment_amount
            FROM `tabAR City Ledger Invoice Payment Entry`
            WHERE payment_entry_id = %s
        """,
            (doc.name,),
            as_dict=True,
        )

        parents = {}
        for r in rows:
            parents.setdefault(r.parent, []).append(r)

        updated = []
        for parent_name, child_rows in parents.items():
            try:
                arci = frappe.get_doc("AR City Ledger Invoice", parent_name)
                total_removed = flt(sum(flt(r.payment_amount) for r in child_rows))

                # remove rows referencing this payment_entry_id
                remaining_rows = [
                    r
                    for r in (arci.get("ar_city_ledger_invoice_payment_entry") or [])
                    if r.get("payment_entry_id") != doc.name
                ]
                arci.ar_city_ledger_invoice_payment_entry = remaining_rows

                # update totals
                arci.total_paid = flt(arci.total_paid) - total_removed
                if arci.total_paid < 0:
                    arci.total_paid = 0.0

                arci.outstanding = flt(arci.outstanding) + total_removed

                # update status
                if flt(arci.total_paid) < flt(arci.total_amount) - 0.0001:
                    arci.status = "Unpaid"

                arci.save(ignore_permissions=True)
                updated.append(parent_name)
            except Exception as e:
                frappe.log_error(
                    _(
                        "Error while removing PE links from ARCI {0} for PE {1}: {2}"
                    ).format(parent_name, doc.name, cstr(e)),
                    _("ARCI Remove PE Link Error"),
                )

        if updated:
            frappe.msgprint(
                _(
                    "Removed Payment Entry {0} allocations from AR City Ledger Invoices: {1}"
                ).format(doc.name, ", ".join(updated))
            )
    except Exception as e:
        frappe.log_error(
            _(
                "Failed to cleanup AR City Ledger Invoice links for Payment Entry {0}: {1}"
            ).format(doc.name, cstr(e)),
            _("ARCI Cleanup Error"),
        )

File: inn/public/js/payment_entry.js

frappe.ui.form.on("Payment Entry", {

    // Handle custom_get_city_ledger_invoice button click (the 'options' value from JSON)
    custom_get_city_ledger_invoice: function (frm) { 
        // Check if customer is selected
        if (!frm.doc.party || frm.doc.party_type !== "Customer") {
            frappe.msgprint(__("Please select a Customer in the 'Party' field and ensure 'Party Type' is 'Customer'."));
            return;
        }
        if (!frm.doc.company) {
            frappe.msgprint(__("Please select a Company first."));
            return;
        }

        frappe.call({
            method: "inn.overrides.payment_entry.get_unpaid_city_ledger_invoices",
            args: {
                customer_id: frm.doc.party, 
                company: frm.doc.company,
            },
            callback: (r) => {
                if (r.message && r.message.length > 0) {
                    frm.clear_table("custom_unpaid_city_ledger_invoice");
                    r.message.forEach(item => {
                        let new_row = frm.add_child("custom_unpaid_city_ledger_invoice");
                        new_row.reference_doctype = item.reference_doctype;
                        new_row.reference_name = item.reference_name;
                        new_row.total_amount = item.total_amount;
                        new_row.outstanding_amount = item.outstanding_amount;
                        new_row.allocated_amount = 0.0;
                    });
                    frm.refresh_field("custom_unpaid_city_ledger_invoice");
                    distribute_paid_amount_to_city_ledger(frm);
                    calculate_city_ledger_allocated_amount(frm);
                } else {
                    frappe.msgprint(__("No unpaid AR City Ledger Invoices found for this customer."), 'blue');
                    frm.clear_table("custom_unpaid_city_ledger_invoice");
                    frm.refresh_field("custom_unpaid_city_ledger_invoice");
                    calculate_city_ledger_allocated_amount(frm);
                }
            }
        });
    },
    
    // Trigger distribution when main Payment Entry's amount changes
    paid_amount: function(frm) {
        distribute_paid_amount_to_city_ledger(frm);
    },
    base_paid_amount: function(frm) {
        distribute_paid_amount_to_city_ledger(frm);
    }
});

frappe.ui.form.on("Unpaid City Ledger Invoice", {
    allocated_amount: function (frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        let allocated = parseFloat(row.allocated_amount) || 0;
        let outstanding = parseFloat(row.outstanding_amount) || 0;

        if (allocated > outstanding) {
            let msg = __("Allocated amount ({0}) for invoice '{1}' cannot exceed its outstanding amount ({2}).");
            msg = msg.replace("{0}", allocated).replace("{1}", row.reference_name).replace("{2}", outstanding);

            frappe.msgprint(msg, __("Over-allocation Warning"));

            frappe.model.set_value(cdt, cdn, 'allocated_amount', outstanding);

            frm.refresh_field('custom_unpaid_city_ledger_invoice');

            let total = calculate_city_ledger_allocated_amount(frm);
            if (frm.doc.hasOwnProperty('custom_total_allocated_city_ledger')) {
                frm.set_value('custom_total_allocated_city_ledger', total);
            }
        } else {
            let total = calculate_city_ledger_allocated_amount(frm);
            if (frm.doc.hasOwnProperty('custom_total_allocated_city_ledger')) {
                frm.set_value('custom_total_allocated_city_ledger', total);
            }
        }
    },
});
// Helper function to calculate total allocated for City Ledger Invoices
function calculate_city_ledger_allocated_amount(frm) {
    let total_allocated = 0;
    if (frm.doc.custom_unpaid_city_ledger_invoice) {
        for (let row of frm.doc.custom_unpaid_city_ledger_invoice) {
            total_allocated += (row.allocated_amount || 0);
        }
    }
    return total_allocated;
}

// Helper function to distribute main Payment Entry's amount to City Ledger Invoices
function distribute_paid_amount_to_city_ledger(frm) {
    let paid_amount = frm.doc.paid_amount || 0;
    let remaining_amount_to_distribute = paid_amount;

    if (!frm.doc.custom_unpaid_city_ledger_invoice || frm.doc.custom_unpaid_city_ledger_invoice.length === 0) {
        return;
    }

    for (let row of frm.doc.custom_unpaid_city_ledger_invoice) {
        let outstanding = row.outstanding_amount || 0;
        let amount_to_allocate = 0;

        if (remaining_amount_to_distribute <= 0) {
            row.allocated_amount = 0;
            continue;
        }

        if (remaining_amount_to_distribute >= outstanding) {
            amount_to_allocate = outstanding;
        } else {
            amount_to_allocate = remaining_amount_to_distribute;
        }

        row.allocated_amount = amount_to_allocate;
        remaining_amount_to_distribute -= amount_to_allocate;
    }
    frm.refresh_field("custom_unpaid_city_ledger_invoice");
}
